packages/caliper-core/lib/worker/rate-control/maxRate.js

```diff
@@ -0,0 +1,203 @@
+                if (this.exceededSampleInterval(resultStats)) {
+                    this.statisticsMap.set('current', resultStats[1]);
+                    this.statisticsMap.set('sampleStart', resultStats[1].final.last);
+                    const achievedTPS = this.retrieveIntervalTPS(resultStats);
+
+                    // New TPS results
+                    this.observedTPS.set('previous', this.observedTPS.get('current'));
+                    this.observedTPS.set('current', achievedTPS);
+
+                    Logger.debug(`Observed current worker TPS ${this.observedTPS.get('current')}`);
+                    Logger.debug(`Observed previous worker TPS ${this.observedTPS.get('previous')}`);
+
+                    // Action based on transaction rate trajectory (+/-)
+                    const dTxn = this.observedTPS.get('current') - this.observedTPS.get('previous');
+                    this.tpsSettings.set('previous', this.tpsSettings.get('current'));
+                    if (dTxn > 0) {
+                        // Keep ramping, try for the new max!
+                        this.tpsSettings.set('current', this.tpsSettings.get('current') + this.step);
+                        Logger.debug(`Increased worker TPS to ${this.tpsSettings.get('current')}`);
+                    } else {
+                        // Too far, back off and try smaller step size. Need to ensure we drain the backlog too.
+                        this.reset = true;
+                        this.tpsSettings.set('current', this.tpsSettings.get('current') - this.step);
+                        this.step = this.step > 0.2 ? this.step / 2 : this.step;
+                        Logger.debug(`Decreased worker TPS to ${this.tpsSettings.get('current')} and step size to ${this.step}`);
+                    }
+                }
+            }
+        }
+
+        // Continue at fixed TPS within this update interval
+        await this.applySleepInterval();
+    }
+
+    /**
+     * Check if a txUpdate has occurred
+     * @param {object} resultStats the result statistics
+     * @returns {boolean} update boolean
+     */
+    updateOccurred(resultStats) {
+        return this.statisticsMap.get('current').create.min !== resultStats[1].create.min;
+    }
+
+    /**
+     * Check if required sample time has been reached
+     * @param {object} resultStats the result statistics
+     * @returns {boolean} boolean flag
+     */
+    exceededSampleInterval(resultStats) {
+        return resultStats[1].final.last - this.statisticsMap.get('sampleStart') >= this.sampleInterval;
+    }
+
+    /**
+     * TPS from the previous txUpdate interval statistics
+     * @param {object} resultStats the passed stats object
+     * @return {number} the TPS within the interval
+     */
+    retrieveIntervalTPS(resultStats) {
+        const resultStatistics = resultStats[1];
+        return (resultStatistics.succ + resultStatistics.fail) / (resultStatistics.final.last - resultStatistics.create.min);
```

Comment:
Shouldn't we only consider the goodput? This way timed out TXs can also contribute to the overall TPS, while they normally indicate an overloaded system. 
Although ignoring expectedly failed TXs could speed up the send rate unintentionally... Insights?

Reply:
hmmm, not sure on this one. My feeling is that if we only account for goodput, we would not be accounting for all the transactions hitting the SUT .. and it might be the case that someone is intentionally wanting to investigate TPS rates with a certain percentage of failed transactions in the mix.

Maybe we need to enable clarification for the intended use of the "count as load" options when related to rate limiting and TPS determination?

Reply:
We could add a rate controller option like `countFailedTxs: true`. And then the user can choose the appropriate setting for the workload.